src/awkward/_v2/_connect/rdataframe/to_rdataframe.py

```diff
@@ -166,16 +177,22 @@ class {array_data_source} final
         }}
 
     public:
-        {array_data_source}(ULong64_t size, std::initializer_list<ULong64_t> ptrs_list)
+        {array_data_source}(PyObject* lookup, ULong64_t size, std::initializer_list<ULong64_t> ptrs_list)
           : fSize(size),
             fPtrs({{ptrs_list}}),
             fColNames({{{cpp_code_column_names}}}),
             fColTypeNames({{{cpp_code_column_type_names}}}),
-            fColTypesMap({{{cpp_code_column_types_map}}})
+            fColTypesMap({{{cpp_code_column_types_map}}}),
+            fPyLookup(lookup)
             {{
+                Py_INCREF(fPyLookup);
                 {cpp_code_init_slots}
             }}
 
+            ~{array_data_source}() {{
+                Py_DECREF(fPyLookup);
+            }}
```

Comment:
Maybe unrelated, but I know that cppyy looks for special attributes on Pythonisations that indicate whether e.g. the GIL can be released. If the GIL is released manually, then I doubt we'd have any introspective mechanism for identifying this, but in the Python->CPPYY case it should be possible to identify:
https://cppyy.readthedocs.io/en/latest/misc.html

Reply:
Indeed, any call to from Python to C++ via cppyy does not release the GIL by default, so the construction should be safe. That default behaviour can be modified on a per-function basis with the `__release_gil__` attribute (as explained in the link that @agoose77 shared).

As @jpivarski pointed out, the destruction does not happen as a result of any explicit call from Python, but it should still hold the GIL since I can't see any explicit release. It happens here:

https://github.com/root-project/root/blob/master/bindings/pyroot/cppyy/CPyCppyy/src/CPPInstance.cxx#L210-L217

One way to double check would be to call `PyGILState_Check()` from the destructor code.